### PR TITLE
implement DoubleEndedIterator, ExactSizeIterator, and FusedIterator for the return value of NonEmpty::iter()

### LINKED ISSUE #37: DoubleEndedIterator?
It would be cool if NonEmpty's iterator type implemented `DoubleEndedIterator`

### PR DESCRIPTION
**Pull request by Roguelazer**
I tried to do this for `iter_mut` too, but it's, um, a bit more complicated.

Fixes #37

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -118,6 +118,47 @@ pub struct NonEmpty<T> {
...
+            Some(first_value)
+        } else {
+            None
+        }
+    }
+}
+
+impl<'a, T> ExactSizeIterator for NonEmptyIter<'a, T> {
+    fn len(&self) -> usize {
+        self.tail.len() + if self.head.is_some() { 1 } else { 0}
```
Comment (FintanH) [inline src/lib.rs:156]:
```suggestion
        self.tail.len() + self.head.map_or(0, |_| 1)
```
Just a small suggestion :)

**Review by FintanH — APPROVED** [review]
This looks good, I just have one small suggestion :)

**Comment by FintanH** [comment]
Thanks for these changes :grin: I'll let @cloudhead have the final call on merging :)

**Comment by cloudhead** [comment]
Looks good. One thing I'm thinking is, perhaps we should call this type simply `Iter`, not to be confused by a non-empty version of Iterator, which is something I'd like to have eventually. So basically it would look like:

`nonempty::Iter` -> Just the name of the iterator for NonEmpty lists
`nonempty::NonEmptyIter` -> An iterator over a NonEmpty, that is never empty

**Comment by Roguelazer** [comment]
Renamed as requested

It would perhaps make sense at some point to move this to `nonempty::vec::{Iter,NonEmpty}` and re-export it from `lib.rs`...

**Comment by cloudhead** [comment]
Thanks!

**Pull request merged**